### PR TITLE
resolve customBuildName with environment

### PR DESCRIPTION
**Pull request by peshay**
In case you want to override the Build Name, it can also be enhanced by environment variables.

### CODE REVIEW

**Comment by peshay** [comment]
https://issues.jenkins-ci.org/plugins/servlet/mobile#issue/JENKINS-46171

**Comment by eyalbe4** [comment]
Thanks for this PR @peshay!
Can you please join our contributors community by signing [JFrog's CLA](https://secure.echosign.com/public/hostedForm?formid=5IYKLZ2RXB543N) so that we can merge your code?

**Comment by peshay** [comment]
@eyalbe4 Actually I already did that. But I just did it again.
And thank you if it can get merged. :)

**Comment by eyalbe4** [comment]
Thanks @peshay.
Which name is the CLA signed with? Your github account includes only *peshay*.

**Comment by peshay** [comment]
Andreas Hubert

File: src/main/java/org/jfrog/hudson/util/BuildUniqueIdentifierHelper.java [hunk 1]
```diff
@@ -139,8 +142,24 @@ public static String getBuildNumber(Run build) {
...
+    
+    public static String getCustomBuildNameResolved(String customBuildName, EnvVars env) {
+    	return Util.replaceMacro(customBuildName, env);
+    }
 
+    public static String getBuildNameConsiderOverride(BuildInfoAwareConfigurator configurator, Run build, EnvVars env) {
+        return configurator.isOverrideBuildName() ? getCustomBuildNameResolved(configurator.getCustomBuildName(), env) : BuildUniqueIdentifierHelper.getBuildName(build);
+    }
+    
     public static String getBuildNameConsiderOverride(BuildInfoAwareConfigurator configurator, Run build) {
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/hudson/util/BuildUniqueIdentifierHelper.java]: I'm concerned about leaving this signature of the *getBuildNameConsiderOverride* method. This could cause bugs, because usages of this APIs may get an unresolved build name.
I believe all usages can be modified to use the new API.
BTW, *build.getEnvironment()* seems to be deprecated.

Comment (peshay) [reply src/main/java/org/jfrog/hudson/util/BuildUniqueIdentifierHelper.java]: As soon as I find the time, I'll try a second attempt. So then I would 
* replace the changed function anywhere where I can find it
* use the TaskListener in order to resolve build environments
* if I'm not able to get the environment map, I would pass null to my function
* in my function I will test if the Override BuildName even has symbols for variables in it
* only if thats the case and I was able to resolve the build environment I will change the build name

**Comment by peshay** [comment]
@eyalbe4 where you able to review my changes? Anything else that I should consider etc.?

**Comment by peshay** [comment]
@eyalbe4 any news or comments to my recent changes?

**Comment by peshay** [comment]
Happy new year @eyalbe4 :) Maybe my Pull Request could be considered? Would be great, so I don't have to remerge and build my own plugin version ;) 
Thanks

File: src/main/java/org/jfrog/hudson/ivy/ArtifactoryIvyRunListener.java [hunk 2]
```diff
@@ -56,12 +56,10 @@ public void onCompleted(AbstractBuild run, TaskListener listener) {
                 return;
             }
             if (artifactoryIvyConfigurator.isDeployBuildInfo()) {
-                String buildName = BuildUniqueIdentifierHelper.getBuildNameConsiderOverride(artifactoryIvyConfigurator, run);
+                String buildName = BuildUniqueIdentifierHelper.getBuildNameConsiderOverride(artifactoryIvyConfigurator, run, null);
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/hudson/ivy/ArtifactoryIvyRunListener.java:59]: Any reason why we sent null here as an argument, instead of env? Would the following work to get the env?
```
EnvVars env = run.getEnvironment(listener);
```

File: src/main/java/org/jfrog/hudson/maven3/ArtifactoryMaven3Configurator.java [hunk 3]
```diff
@@ -394,7 +395,13 @@ public void buildEnvVars(Map<String, String> env) {
             public boolean tearDown(AbstractBuild build, BuildListener listener) {
                 Result result = build.getResult();
                 if (deployBuildInfo && result != null && result.isBetterOrEqualTo(Result.SUCCESS)) {
-                    String buildName = BuildUniqueIdentifierHelper.getBuildNameConsiderOverride(ArtifactoryMaven3Configurator.this, build);
+                    EnvVars env;
+                    try {
+                        env = build.getEnvironment(listener);
+                    } catch (Exception e) {
+                        env = null;
+                    }
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/hudson/maven3/ArtifactoryMaven3Configurator.java:403]: It looks like here you chose to wrap the ```build.getEnvironment```  call with try catch, unlike in other places where you used this API.
Why is that?

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
See my inline comments.
I think the last issue is what concerns me the most - The link to the build-info gets broken if env vars are used.
Please let me know what you think.

**Comment by eyalbe4** [comment]
@peshay,
Are you planning to continue working on this?

**Comment by peshay** [comment]
@eyalbe4 I am not a full time developer, so it's hard for me to implement your suggestions. It would take some time for me to work into this topic again. I would hope someone else could finish this task for me with more development experience. Still this is a functionally is one I need and I patch this branch into my Jenkins Projects for a while. 
tldr: If there won't be anyone else implementing such a feature, I may work on it again in a few months or so ...
